scipy/spatial/_hausdorff.pyx

```diff
@@ -16,56 +16,73 @@ import numpy as np
 cimport numpy as np
 cimport cython
 from libc.math cimport sqrt
+from libc.stdlib cimport srand, rand, RAND_MAX
+from libc.limits cimport UINT_MAX
 
 __all__ = ['directed_hausdorff']
 
+cdef extern from "numpy/npy_math.h":
+    double inf "NPY_INFINITY"
+
+@cython.wraparound(False)
+@cython.boundscheck(False)
+@cython.cdivision(True)
+cdef void index_shuffler(long [::1] arr, size_t size):
```

Comment:
Writing robust C-level array shufflers is non-trivial. If we use this, the [source](https://benpfaff.org/writings/clc/shuffle.html) should be attributed.

The primary reason for doing this, in the second commit in this PR, was to alleviate the issue of the N=10 benchmark actually slowing down.

I'd prefer to use the Cython implementation, especially if randomgen eventually exposes the machinery I need with little overhead (@mattip / @bashtage ?), but for the sake of demonstration, we can make all currently-benchmarked data sizes faster using this.

Keep in mind though, C `rand()` has its weaknesses as the linked algorithm notes, and even though our `asv` benchmarks now show pretty big improvements across the board, the average performance on massive and structured data sets may fluctuate depending on `rand()` quality.

Reply:
Are you looking for the [shuffle function](https://github.com/mattip/numpy/blob/randomgen/numpy/random/randomgen/generator.pyx#L3999) from the new randomgen branch?

Reply:
If you're concerned about the quality of `rand` then you might want to compile with C++ and look at their PRNGs. There's even a `shuffle` algorithm that you could use here. You will need to give cython the declarations manually:
```cython
from libc.stdint import uint_fast32_t

cdef extern from "<random>" namespace "std" nogil:
    cdef cppclass minstd_rand:
        minstd_rand()
        minstd_rand(uint_fast32_t)

cdef extern from "<algorithm>" namespace "std" nogil:
    void shuffle[Iter, RNG](Iter first, Iter last, RNG& rng)
```

And then it should be usable like this:
```
cdef minstd_rand rng = minstd_rand(seed)
cdef long[::1] arr = np.arange(size, dtype=np.long)
shuffle(&arr[0], &arr[0] + size, rng)
```

`minstd_rand` is an LCG so possibly not much better than `rand` but is at least consistent in quality across platforms (`rand` is only 15 bits on windows!). There is also a mersenne twister engine but constructing a full mersenne twister for each call to `directed_hausdorff` would probably be excessive.